### PR TITLE
feat(instance): move ip attachement to server

### LINKED ISSUE #349: Can't attach an IP to a server
Currently the only way to attach an IP to a server is to set the server ID to the IP. It would be nice to be able to give an IP ID to the server also

### PR DESCRIPTION
**Pull request by Sh4d1**
Fixes #349 

Signed-off-by: Patrik Cyvoct <pcyvoct@scaleway.com>

### CODE REVIEW

File: scaleway/resource_instance_server.go [hunk 1]
```diff
@@ -505,6 +521,32 @@ func resourceScalewayInstanceServerUpdate(d *schema.ResourceData, m interface{})
...
+		server, err := instanceAPI.GetServer(&instance.GetServerRequest{
+			Zone:     zone,
+			ServerID: ID,
+		})
+		if err != nil {
+			return err
+		}
+		newIPID := d.Get("ip_id")
+		if server.Server.PublicIP != nil && server.Server.PublicIP.ID != expandID(newIPID) {
+			if newIPID.(string) != "" {
```
Comment (jerome-quere) [inline scaleway/resource_instance_server.go]: missing something like:
```
If server.Server.PublicIP.ID != nil {
  Detach IP
}
```

File: scaleway/resource_instance_server_test.go [hunk 2]
```diff
@@ -428,6 +428,25 @@ func TestAccScalewayInstanceServerIpv6(t *testing.T) {
...
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCheckScalewayInstanceServerConfigWithReservedIP,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckScalewayInstanceServerExists("scaleway_instance_server.base"),
+					testAccCheckScalewayInstanceIPExists("scaleway_instance_ip.ip"),
+					resource.TestCheckResourceAttrPair("scaleway_instance_ip.ip", "address", "scaleway_instance_server.base", "public_ip"),
+					resource.TestCheckResourceAttrPair("scaleway_instance_ip.ip", "id", "scaleway_instance_server.base", "ip_id"),
+				),
+			},
```
Comment (jerome-quere) [inline scaleway/resource_instance_server_test.go]: Add a step where you change an IP by another.  This is the point not cover by your code see comment above

File: website/docs/r/instance_ip.html.markdown [hunk 3]
```diff
@@ -20,7 +20,7 @@ resource "scaleway_instance_ip" "server_ip" {}
 The following arguments are supported:
 
 - `reverse` - (Deprecated) Please use the `scaleway_instance_ip_reverse_dns` resource instead.
-- `server_id` - (Optional) The ID of the server you want to attach this resource to.
+- `server_id` - (Deprecated) The ID of the server you want to attach this resource to.
```
Comment (jerome-quere) [inline website/docs/r/instance_ip.html.markdown]: Add `Deprecated in favor of scaleway_instance_server.ip_id`
If possible add a lik to the correct page on ` scaleway_instance_server.ip_id`

File: scaleway/resource_instance_server.go (hunk 1 above)
Comment (Sh4d1) [reply scaleway/resource_instance_server.go]: Hmm no need, an attach will detach the old IP

File: scaleway/resource_instance_server_test.go (hunk 2 above)
Comment (Sh4d1) [reply scaleway/resource_instance_server_test.go]: :+1:

File: website/docs/r/instance_ip.html.markdown (hunk 3 above)
Comment (Sh4d1) [reply website/docs/r/instance_ip.html.markdown]: :+1:

File: scaleway/resource_instance_server.go (hunk 1 above)
Comment (QuentinBrosse) [reply scaleway/resource_instance_server.go]: I agree with @jerome-quere, if user wants to remove the public IP he can't. We need to detach IP if `ip_id` was not empty and it now empty.

Comment (Sh4d1) [reply scaleway/resource_instance_server.go]: Yep agree, didn't see it like this

Comment (Sh4d1) [reply scaleway/resource_instance_server.go]: fixed

**Comment by Sh4d1** [comment]
@QuentinBrosse @jerome-quere  all good!

**Comment by Sh4d1** [comment]
@kindermoumoute works now

File: scaleway/resource_instance_server.go [hunk 4]
```diff
@@ -505,6 +521,48 @@ func resourceScalewayInstanceServerUpdate(d *schema.ResourceData, m interface{})
...
+	////
+	if d.HasChange("ip_id") {
+		server, err := instanceAPI.GetServer(&instance.GetServerRequest{
+			Zone:     zone,
+			ServerID: ID,
+		})
+		if err != nil {
+			return err
+		}
+		newIPID := d.Get("ip_id")
```
Comment (kindermoumoute) [inline scaleway/resource_instance_server.go]: `newIPID := expandID(d.Get("ip_id"))`

File: scaleway/resource_instance_server.go [hunk 5]
```diff
@@ -505,6 +521,48 @@ func resourceScalewayInstanceServerUpdate(d *schema.ResourceData, m interface{})
...
+		if server.Server.PublicIP == nil && expandID(newIPID) != "" {
```
Comment (kindermoumoute) [inline scaleway/resource_instance_server.go]: remove expandID from here

File: scaleway/resource_instance_server.go [hunk 6]
```diff
@@ -505,6 +521,48 @@ func resourceScalewayInstanceServerUpdate(d *schema.ResourceData, m interface{})
...
+			_, err = instanceAPI.AttachIP(&instance.AttachIPRequest{
+				Zone:     zone,
+				IP:       expandID(newIPID.(string)),
```
Comment (kindermoumoute) [inline scaleway/resource_instance_server.go]: remove expandID from here

File: scaleway/resource_instance_server.go [hunk 7]
```diff
@@ -505,6 +521,48 @@ func resourceScalewayInstanceServerUpdate(d *schema.ResourceData, m interface{})
...
+				ServerID: ID,
+			})
+			if err != nil {
+				return err
+			}
+		} else if server.Server.PublicIP != nil && server.Server.PublicIP.ID != expandID(newIPID) {
```
Comment (kindermoumoute) [inline scaleway/resource_instance_server.go]: remove expandID from here

File: scaleway/resource_instance_server.go [hunk 8]
```diff
@@ -505,6 +521,48 @@ func resourceScalewayInstanceServerUpdate(d *schema.ResourceData, m interface{})
...
+			_, err = instanceAPI.DetachIP(&instance.DetachIPRequest{
+				Zone: zone,
+				IP:   server.Server.PublicIP.ID,
+			})
+			if err != nil {
+				return err
+			}
+			if expandID(newIPID) != "" {
```
Comment (kindermoumoute) [inline scaleway/resource_instance_server.go]: remove expandID from here

File: scaleway/resource_instance_server.go [hunk 9]
```diff
@@ -505,6 +521,48 @@ func resourceScalewayInstanceServerUpdate(d *schema.ResourceData, m interface{})
...
+				_, err = instanceAPI.AttachIP(&instance.AttachIPRequest{
+					Zone:     zone,
+					IP:       expandID(newIPID.(string)),
```
Comment (kindermoumoute) [inline scaleway/resource_instance_server.go]: remove expandID from here

**Pull request merged**